### PR TITLE
SignedEciesCiphertext Support

### PR DESCRIPTION
**Pull request by neekolas**
## Summary

- Removes dependency on `bitchan/eccrypto` and replaces it with a Typescript port that does not have any dependencies on Node crypto and does not require polyfills
- Added support for the new `SignedEciesCiphertext` proto type
- Updated `EncryptedPersistence` to use `SignedEciesCiphertext` and the new `ecies` crypto library

## Notes

- I ported eccrypto to Typescript as best as I could so that it matches the rest of the repo, satisfies the lint rules, and gains some type safety. The type checker is happy, but it's totally possible I misinterpreted some `Uint8Arrays` as `Buffers` or vice versa. If I made any mistake of that category, it wouldn't matter at runtime. Just something to be aware of.
- eccrypto uses a different `secp256k1` library than we do. In a dream world, we would carefully replace all `elliptic` usage with `@noble/secp256k1` or a Rust-based implementation. That is a scarier project than I care to take on today, but it's something we should consider to reduce library bloat.
- I've highlighted all the changes I had to make to `eccrypto` other than type annotations
- Given that `eccrypto` is basically unmaintained and we already had to make significant changes, I have not prioritized the ability to easily merge upstream updates into our vendored version. The move to Typescript would make automated merging difficult anyways. If they happen to release some important security update, we'll have to manually merge it into our version.

### CODE REVIEW

File: src/crypto/ecies.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  for (let i = 0; i < b1.length; i++) {
+    res |= b1[i] ^ b2[i] // jshint ignore:line
+  }
+  return res === 0
+}
+
+/* This must check if we're in the browser or
+not, since the functions are different and does
+not convert using browserify */
+function randomBytes(size: number): Buffer {
```
Comment (neekolas) [inline src/crypto/ecies.ts:56]: Replaces this https://github.com/bitchan/eccrypto/blob/master/browser.js#L48

File: src/crypto/ecies.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  const arr = new Uint8Array(size)
+  browserCrypto.getRandomValues(arr)
+  return Buffer.from(arr)
+}
+
+async function sha512(msg: Buffer) {
```
Comment (neekolas) [inline src/crypto/ecies.ts:62]: Replaces [this](https://github.com/bitchan/eccrypto/blob/master/browser.js#L58)

File: src/crypto/ecies.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+          resolve(Buffer.from(new Uint8Array(result)))
+        })
+    })
+  }
+}
+
+const aesCbcEncrypt = getAes('encrypt')
+const aesCbcDecrypt = getAes('decrypt')
+
+async function hmacSha256Sign(key: Buffer, msg: Buffer) {
```
Comment (neekolas) [inline src/crypto/ecies.ts:89]: Replaces [this](https://github.com/bitchan/eccrypto/blob/master/browser.js#L99)

File: src/crypto/ecies.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+// function hmacSha256Sign(key, msg) {
+//   return new Promise(function (resolve) {
+//     const hmac = nodeCrypto.createHmac('sha256', Buffer.from(key))
+//     hmac.update(msg)
+//     const result = hmac.digest()
+//     resolve(result)
+//   })
+// }
+
+async function hmacSha256Verify(key: Buffer, msg: Buffer, sig: Buffer) {
```
Comment (neekolas) [inline src/crypto/ecies.ts:112]: Replaces [this](https://github.com/bitchan/eccrypto/blob/master/browser.js#L108)

File: src/crypto/ecies.ts [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+      assert(publicKeyB[0] === 2 || publicKeyB[0] === 3, 'Bad public key')
+    }
+    const keyA = ec.keyFromPrivate(privateKeyA)
+    const keyB = ec.keyFromPublic(publicKeyB)
+    const Px = keyA.derive(keyB.getPublic()) // BN instance
+    resolve(Buffer.from(Px.toArray()))
+  })
+}
+
+export async function encrypt(
```
Comment (neekolas) [inline src/crypto/ecies.ts:215]: The eccrypto version was difficult to read with a chain of promises. I converted to async/await, but the logic remains identical (save for the `isValidPrivateKey` bug noted below)

File: src/crypto/ecies.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  msg: Buffer,
+  opts?: { ephemPrivateKey?: Buffer; iv?: Buffer } | undefined
+) {
+  opts = opts || {}
+  // Take IV from opts or generate randomly
+  const iv = opts?.iv || randomBytes(16)
+  let ephemPrivateKey = opts?.ephemPrivateKey || randomBytes(32)
+  // There is a very unlikely possibility that it is not a valid key
+  while (!isValidPrivateKey(ephemPrivateKey)) {
+    if (opts?.ephemPrivateKey) {
```
Comment (neekolas) [inline src/crypto/ecies.ts:226]: I added this if statement and error throwing. This was a bug in `eccrypto`. If an invalid `opts.ephemPrivateKey` is provided, the while loop would run infinitely in [their implementation](https://github.com/bitchan/eccrypto/blob/master/browser.js#L216).

File: src/crypto/ecies.ts [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+  // Return the payload
+  return {
+    iv,
+    ephemeralPublicKey,
+    ciphertext,
+    mac,
+  }
+}
+
+export async function decrypt(privateKey: Buffer, opts: Ecies) {
```
Comment (neekolas) [inline src/crypto/ecies.ts:252]: Also converted to `async/await` from [original](https://github.com/bitchan/eccrypto/blob/master/browser.js#L243) for readability

File: test/keystore/InMemoryKeystore.test.ts [hunk 8]
```diff
@@ -473,6 +475,27 @@ describe('InMemoryKeystore', () => {
     it('works with persistence', async () => {})
   })
 
+  describe('createAuthToken', () => {
```
Comment (neekolas) [inline test/keystore/InMemoryKeystore.test.ts:478]: This test got misplaced in the splitting of the other two PRs.

File: src/keystore/persistence/EncryptedPersistence.ts [hunk 9]
```diff
@@ -94,13 +41,35 @@ export default class EncryptedPersistence implements Persistence {
...
+  private async deserializeEcies(data: Uint8Array): Promise<Ecies> {
+    const protoVal = SignedEciesCiphertext.fromBytes(data)
+    // Verify the signature upon deserializing
+    if (!(await protoVal.verify(this.privateKey.publicKey))) {
+      throw new Error('signature validation failed')
+    }
+    const ecies = protoVal.ciphertext
+
+    return {
+      ciphertext: Buffer.from(ecies.ciphertext),
```
Comment (neekolas) [inline src/keystore/persistence/EncryptedPersistence.ts:69]: The `ecies` library expects Buffers, and our protos deal in `Uint8Array`s. At some point we can decide whether we want to update `ecies` to take `Uint8Array` but for now I just convert everything.

File: src/crypto/ecies.ts [hunk 10]
```diff
@@ -0,0 +1,253 @@
+// This file is taken from `bitchan/eccrypto` and ported to TS. All references to `nodeCrypto` have been replaced with `browserCrypto`
+import { ec as EC } from 'elliptic'
+import { crypto as browserCrypto } from './encryption'
```
Comment (neekolas) [inline src/crypto/ecies.ts:3]: Used our own resolution of `webcrypto`

File: src/crypto/ecies.ts (hunk 6 above)
Comment (michaelx11) [reply src/crypto/ecies.ts:214]: Wow nice find. Kind of crazy to see that while loop just repeatedly reassigning `ephemPrivateKey = opts?.ephemPrivateKey` in the case it's specified.

File: src/crypto/ecies.ts (hunk 7 above)
Comment (michaelx11) [reply src/crypto/ecies.ts:240]: At this point you might as well ship your own open source fork of eccrypto haha

File: src/keystore/InviteStore.ts [hunk 11]
```diff
@@ -4,7 +4,7 @@ import type { TopicData } from './interfaces'
 import { Mutex } from 'async-mutex'
 import { typeSafeTopicMap } from './utils'
 
-const INVITE_KEY = 'invitations:v1'
+const INVITE_KEY = 'invitations/v1'
```
Comment (michaelx11) [inline src/keystore/InviteStore.ts:7]: This changes seems a bit unrelated, but shouldn't have an impact right? No internal keystore state convo state is persisted yet.

File: test/keystore/persistence/EncryptedPersistence.test.ts [hunk 12]
```diff
@@ -100,34 +98,43 @@ describe('EncryptedPersistence', () => {
...
+        await PrivateKeyBundleV1.generate()
+      ).identityKey
+    )
+    await persistence.setItem(TEST_KEY, signedBySomeoneElse.toBytes())
 
-    // Expect an error if the ciphertext is tampered with
-    await expect(encryptedPersistence.getItem(TEST_KEY)).rejects.toThrow(
-      'Invalid data length'
+    expect(encryptedPersistence.getItem(TEST_KEY)).rejects.toThrow(
+      'signature validation failed'
```
Comment (michaelx11) [inline test/keystore/persistence/EncryptedPersistence.test.ts:137]: nice conversion! Is it still worth having a "signed correctly, but encrypted incorrectly" test case?

**Review by michaelx11 — APPROVED** [review]
LGTM - super tricky to validate but since it's a 1:1 port I'm okay with doing a rough inspection for now

File: src/keystore/InviteStore.ts (hunk 11 above)
Comment (neekolas) [reply src/keystore/InviteStore.ts:7]: That's right. Just wanted to make the suffix consistent with the prefix (switched to slashes to make a potential `FileSystemPersistence` module prettier)

File: test/keystore/persistence/EncryptedPersistence.test.ts (hunk 12 above)
Comment (neekolas) [reply test/keystore/persistence/EncryptedPersistence.test.ts:137]: Good call. Will add.

Comment (neekolas) [reply test/keystore/persistence/EncryptedPersistence.test.ts:138]: Added

**Pull request merged**